tests/test_jaxpr_jvp.py

```diff
@@ -0,0 +1,448 @@
+
+    @pytest.mark.parametrize("use_jit", (False, True))
+    def test_basic_circuit(self, use_jit):
+        """Test the calculation of results and jvp for a basic circuit."""
+
+        def f(x):
+            qml.RX(x, 0)
+            return qml.expval(qml.Z(0))
+
+        jaxpr = jax.make_jaxpr(f)(0.5)
+
+        args = (0.82,)
+        tangents = (2.0,)
+        dev_jaxpr_jvp = qml.device(device_name, wires=1).jaxpr_jvp
+
+        executor = partial(dev_jaxpr_jvp, jaxpr.jaxpr)
+        if use_jit:
+            executor = jax.jit(executor)
+
+        results, dresults = executor(args, tangents)
+
+        assert len(results) == 1
+        assert qml.math.allclose(results, jax.numpy.cos(args[0]))
+        assert len(dresults) == 1
+        assert qml.math.allclose(dresults[0], tangents[0] * -jax.numpy.sin(args[0]))
+
+    def test_diffentiable_op_math(self):
+        """Test that we can handle differentiable operations in the circuit."""
+
+        def f(x):
+            qml.adjoint(qml.RX(x, 0))
+            return qml.expval(qml.Z(0))
+
+        jaxpr = jax.make_jaxpr(f)(0.5)
+
+        args = (0.82,)
+        tangents = (2.0,)
+        dev_jaxpr_jvp = qml.device(device_name, wires=1).jaxpr_jvp
+
+        results, dresults = dev_jaxpr_jvp(jaxpr.jaxpr, args, tangents)
+
+        assert len(results) == 1
+        assert qml.math.allclose(results, jax.numpy.cos(args[0]))
+        assert len(dresults) == 1
+        assert qml.math.allclose(dresults[0], tangents[0] * -jax.numpy.sin(args[0]))
+
+    def test_abstract_zero_tangent(self):
+        """Test we get the derivatives will be zero if the tangent is abstract zero."""
+
+        def f(x):
+            _ = x + 1
+            qml.RX(0.5, 0)
+            return qml.expval(qml.Z(0))
+
+        args = (0.5,)
+        tangents = (jax.interpreters.ad.Zero(jax.core.ShapedArray((), float)),)
+
+        jaxpr = jax.make_jaxpr(f)(0.5)
+        [results], [dresults] = qml.device(device_name, wires=1).jaxpr_jvp(
+            jaxpr.jaxpr, args, tangents
+        )
+
+        assert qml.math.allclose(results, jax.numpy.cos(0.5))
+        assert qml.math.allclose(dresults, 0)
+
+    def test_multiple_in(self):
+        """Test that we can differentiate multiple inputs."""
+
+        def f(x, y):
+            qml.RX(x, 0)
+            qml.RY(y, 1)
+            qml.CNOT((0, 1))
+            return qml.expval(qml.Y(0))
+
+        x = jax.numpy.array(0.5)
+        y = jax.numpy.array(1.2)
+        dx = jax.numpy.array(2.0)
+        dy = jax.numpy.array(3.0)
+
+        jaxpr = jax.make_jaxpr(f)(x, y)
+
+        dev_jaxpr_jvp = qml.device(device_name, wires=2).jaxpr_jvp
+
+        [res], [dres] = dev_jaxpr_jvp(jaxpr.jaxpr, (x, y), (dx, dy))
+        expected = -jax.numpy.sin(x) * jax.numpy.sin(y)
+        assert qml.math.allclose(res, expected)
+
+        expected_dres = dx * -jax.numpy.cos(x) * jax.numpy.sin(y) + dy * -jax.numpy.sin(
+            x
+        ) * jax.numpy.cos(y)
```

Comment:
```suggestion
        expected_dres1 = dx * -jax.numpy.cos(x) * jax.numpy.sin(y)
        expected_dres2 = dy * -jax.numpy.sin(x) * jax.numpy.cos(y)
        expected_dres = expected_dres1 + expected_dres2
```
or anything else that makes this readable :)